10m populated places fix wikidataids

mizmay opened the pull request:
Connects to  #419, #453,  #449, #411, #406, #339, #405, #373, #365, #358, #356, #334, #330, #323, #318, #389, #388, #387, #385, #384, #383, #381, #377, #371, #368, #367, #364, #362, #310, #293.

This PR partially or fully addresses over 30 issues. The goal here is to make the changes necessary in advance of batch updates to default and localized names. The only really necessary change I know of is to update wikidataids and add new features, but since updating `FEATURECLA` can happen independently, I addressed issues around changes to designation as `Populated place`, `Admin-1 capital` or `Admin-0 capital` as well.

Modifications to 10m_Populated_Places in keeping with Step 1 and Step 2, outlined in [this gist](https://gist.github.com/mizmay/89f6eed0e571e5e324bb7e13005715fc). 

An additional large chunk of open issues (25+) can be addressed once localized name_* values are updated based on wikidataids, leaving the path clear for additional batch processes and/or manual edits to individual records.

1. Script documents changes to `wikidataid`s to support overwriting default `NAME`, `NAMEASCII`, and localized `name_*` fields based on values from wikidata.
2. CSV lists the full attributes for the records with revised `wikidataid`

curran commented:
I notice that CI is broken starting at commit https://github.com/nvkelso/natural-earth-vector/pull/457/commits/0d14e636248571a6000333d730817ca823ad11c5

curran commented:
Oh wait, sorry, CI is broken in `master` anyway, so I guess it's a non-issue.

nvkelso commented:
Yep, I need to investigate why. Do you see anything obvious in logs?


> On Jan 5, 2021, at 10:17, Curran Kelleher <notifications@github.com> wrote:
> 
> ﻿
> Oh wait, sorry, CI is broken in master anyway, so I guess it's a non-issue.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

curran commented:
The breakage was introduced in the last commit it seems.

![image](https://user-images.githubusercontent.com/68416/103705243-afdb8080-4f78-11eb-961c-89e2624740bf.png)

curran commented:
But that doesn't make sense, that change was trivial https://github.com/nvkelso/natural-earth-vector/commit/bf7720b54dd9ac2d4d7f735174901b3862b5362a

Maybe there are some dependencies or something whose versions are not locked? No idea... Maybe just technology churn.

curran commented:
It fails on `brew uninstall python mercurial postgis sfcgal cgal gdal`

https://travis-ci.org/github/nvkelso/natural-earth-vector/jobs/749392272

mizmay commented:
That makes sense! It’s broken on all commits for all PRs.

Sent from my iPhone

> On Jan 5, 2021, at 2:11 PM, Curran Kelleher <notifications@github.com> wrote:
> 
> ﻿
> It fails on brew uninstall python mercurial postgis sfcgal cgal gdal
> 
> https://travis-ci.org/github/nvkelso/natural-earth-vector/jobs/749392272
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

nvkelso commented:
Let's track the CI issue separately in https://github.com/nvkelso/natural-earth-vector/issues/464.

nvkelso commented:
Can you include the script(s) that generate these two CSVs, please?

- Comparison between suggestions and Wikidata
- All label/name fields that could be adjusted to match Wikidata value

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  mutate(across(where(is.factor), as.character)) %>%
+  #453 Rida -> Rada'a
+  rows_update(by = "ne_id", tibble(ne_id = 1159134263,
+                                   wikidataid = 'Q2125362')) %>%
```
The default **name** value should change to `Rada'a` and **name_alt** to `Rida` to preserve the earlier value, and **nameascii** value should change to `Radaa`. The **geonameid** should be `71491`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   wikidataid = 'Q2125362')) %>%
+  #449 Sakarya
+  rows_update(by = "ne_id", tibble(ne_id = 1159113425, 
+                                   wikidataid = 'Q3945164')) %>%
```
**Adapazarı** with ne_id of `1159146375 ` should be changed as **featurecla** to `Admin-1 capital`. The other place near here **Sakarya** should be deleted. Technically both these places are true in their own ways, but we only need a single point here for balance in the overall composition. 

The **name_alt** of `1159146375` should change to `Sakarya`.

Delete Sakarya **ne_id** `1159113425` entirely.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   wikidataid = 'Q3945164')) %>%
+  #419 Saguenay
+  rows_update(by = "ne_id", tibble(ne_id = 1159148923,
+                                   wikidataid = 'Q139229')) %>%
```
For **ne_id** `1159148923` , the default **name** value should change to `Saguenay` and **name_alt** to `Chicoutimi` to preserve the earlier value, and **nameascii** value should change to `Saguenay`. The GeoNames ID is good as-is. The **pop_max** is `144746` and the new **pop_rank** is `9`.

mizmay replied:
Modified POP_MAX and RANK_MAX but not POP_MIN or RANK_MIN

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   wikidataid = 'Q139229')) %>%
+  #411 Jinxi -> Huludao
+  rows_update(by = "ne_id", tibble(ne_id = 1159149137,
+                                   wikidataid = 'Q75379')) %>%
```
For **ne_id** `1159149137` , the default **name** value should change to `Huludao ` and **name_alt** to `Jinxi` to preserve the earlier value, and **nameascii** value should change to `Huludao`. The **pop_max** is `724800` and the new **pop_rank** is `11`.

mizmay replied:
Modified POP_MAX and RANK_MAX but not POP_MIN or RANK_MIN

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ADM1NAME = 'Anambra',
+                                   ISO_A2 = 'NG',
+                                   TIMEZONE = 'Africa/Lagos',
+                                   LATITUDE = 6.0195310,
```
Don't know that this also sets the geometry to the lat/lng.

nvkelso replied:
All new features need a **ne_id** added. Those are generated from https://www.brooklynintegers.com/create/.

Please also update: **ne_id** to `1729764357`.

mizmay replied:
It does not. Geometry needs to be added in a separate step.

Modified POP_MAX and RANK_MAX but not POP_MIN or RANK_MIN.

nvkelso replied:
NOTE: I reset a few of the `POP_MIN`, `POP_OTHER`, `RANK_MIN` changes. There's some logic around which is the "core" population value and the "metro-area" population value that can be considered, but let's ignore that in this PR.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   LATITUDE = 6.0195310,
+                                   LONGITUDE = 6.9171810)) %>%
+  #406 Bose -> Baise
+  rows_update(by = "ne_id", tibble(ne_id = 1159149783,
```
For **ne_id** `1159149783` , the default **name** value should change to `Baise` and **name_alt** to `Bose` to preserve the earlier value, and **nameascii** value should change to `Baise`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159149783,
+                                   wikidataid = 'Q571949')) %>%
+  #405 Remove North Shore, NZ
+  rows_delete(by = "ne_id", tibble(ne_id = 1159149783)) %>%
```
Agree with this, but add a new line for white space.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_delete(by = "ne_id", tibble(ne_id = 1159149783)) %>%
+  #389 Dar es Salaam should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151305,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
I disagree with this.

Instead the **FEATURECLA** should be `Admin-0 capital alt`. There are still national gov't operations here.

But it must be paired with **ne_id** of `1159149731` to **FEATURECLA** of `Admin-0 capital`. The official capital has been moving here from the 1970s... but it's been slow going but seems mostly completed.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #388 Kyoto should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159149967,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
I disagree with this.

I suppose we can make all the **FCLASS** variants `Admin-1 capital`... but in Natural Earth-isms the default **FEATURELCA** should not be changed from `Admin-0 capital alt`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #387 Tomakomai should be Populated
+  rows_update(by = "ne_id", tibble(ne_id = 1159141281,
+                                   FEATURECLA = 'Populated place')) %>%
```
I agree with this.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Populated place')) %>%
+  #385 Lagos should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151591,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
I agree with this.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #384 Zhengzhou should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151375,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
I agree with this change.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #383 Shijiazhuang should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159149907,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
I agree with this change...

But there are additional changes to make.

For **ne_id** `1159149907 ` , the default **name** value should change to `Shijiazhuang` and **name_alt** to `Shimen, Shihkiachwang` to preserve the earlier value, and **nameascii** value should change to `Shijiazhuang`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #381 Yangon should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151477,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
Let's go a little slower here. While most administrative functions _have_ moved to the new capital, many of the ministry's functions still happen in the old capital and many foreign embassies are still in the old capital. 

For **ne_id** `1159151477`, **FEATURECLA** to 'Admin-1 capital alt', the default **name** value should change to `Yangon` and **name_alt** to `Rangoon` to preserve the earlier value, and **nameascii** value should change to `Yangon`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  #381 Yangon should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151477,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #377 Yangon should be Admin-1 capital
```
The Issue number is correct, but the copy-paste text should be updated to `Makkah (Saudi Arabia) should be Admin-1 capital`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #377 Yangon should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151279,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #373 Remove Vila Velha (Amapá, Brazil)
+  rows_delete(by = "ne_id", tibble(ne_id = 1159148899)) %>%
+  #371 Macau should be Admin-1 capital
```
Add white space between `rows_delete`... and `#371...`.

nvkelso replied:
Please remove this change, it's not in keeping with the rest of the data and I've closed the issue.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159151279,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #373 Remove Vila Velha (Amapá, Brazil)
+  rows_delete(by = "ne_id", tibble(ne_id = 1159148899)) %>%
```
This is a small regional center, but with bunk population data that should be updated:

**wikidataid** of `Q18471012` and set **pop_max**, **pop_min**, **pop_other** to `2723` with new **rank_min** and **rank_max** of `4` and set **min_zoom** to `7.0` to demote it a bit.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  #371 Macau should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159149085,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #371 Hong Kong should be Admin-1 capital
```
Same as Macao above, for Hong Kong please remove this change, it's not in keeping with the rest of the data and I've closed the issue.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #368 Patra should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159147263,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #367 Novi Sad should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159144059,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
🚫 Please remove this.

The current **featurecla** of `Admin-0 region capital` reflects it's status as capital of the admin-0 region of `Autonomous Province of Vojvodina`, one of the two regions of Serbia that sit above the province level. Often region capitals are also province capitals, but the higher order wins for the single feature class status.

LukasGelbmann replied:
Does that mean `adm0name` of Novi Sad should be 'Autonomous Province of Vojvodina' instead of 'Serbia'?

nvkelso replied:
Good question! No, in this case the country should still be `Serbia`. 

The names of these admin-0 "region" subdivisions change per country ("constituent countries" versus "autonomous province") but same concept.  For example: in the `United Kingdom` we also see `Admin-0 region capital` feature classes for **Edinburgh** (Scotland), **Cardiff** (Wales), and **Belfast** (Northern Ireland) where the subdivision (Scottland, Wales, and Northern Ireland) is below the admin-0 country level but above the "province" admin-1 level – and London is seat of national gov't so gets the more overarching `Admin-0 capital`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  #365 Remove duplicate Bandar Lampung (Sumatra, Indonesia)
+  rows_delete(by = "ne_id", tibble(ne_id = 1159139819)) %>%
+  rows_update(by = "ne_id", tibble(ne_id = 1159149817,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159144059,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #365 Remove duplicate Bandar Lampung (Sumatra, Indonesia)
+  rows_delete(by = "ne_id", tibble(ne_id = 1159139819)) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #364 Johannesburg should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151515,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
⛔ No, this should be `Admin-0 capital alt` instead.

Johannesburg is the seat of the [Constitutional Court](https://en.wikipedia.org/wiki/Constitutional_Court_of_South_Africa), the highest court in South Africa so under Natural Earth's peculiar system of measure, it's a capital but (like Valparaíso in Chile) should instead be marked `Admin-0 capital alt`.

This is in addition to the formal judicial capital at [https://en.wikipedia.org/wiki/Bloemfontein](Bloemfontein) which is HQ to [https://en.wikipedia.org/wiki/Supreme_Court_of_Appeal_(South_Africa)](Supreme_Court_of_Appeal_) which hears non-constitutional matters.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #362 Xian should be Admin-1 capital
+  rows_update(by = "ne_id", tibble(ne_id = 1159151363,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+write.csv(ne_10m_populated_places_modified_metadata, "ne_10m_populated_places_modified_metadata.csv", na = "")
+
+# Requires a bit more tweaking to ensure DBF encoding and data types aren't messed up in the transformation.
+#ne_10m_populated_places_modified <- st_set_geometry(ne_10m_populated_places_modified_metadata, 
```
Good luck sort out exporting as SHP with correct DBF encoding, let's hold the PR on that.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+#                     ne_10m_populated_places_modified_metadata)
+# 
+# 
+# write.csv(add_modify, "Add_Modify_10m_Populated_Places_WikidataIDs.csv", na = "")
```
These summaries are useful to put in the PR, thank you!

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159151363,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  #358 Remove Natal in Amazonas, Brazil
+  rows_delete(by = "ne_id", tibble(ne_id = 1159148743)) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  #358 Remove Natal in Amazonas, Brazil
+  rows_delete(by = "ne_id", tibble(ne_id = 1159148743)) %>%
+  #356 Remove Noginsk in the Moscow region
+  rows_delete(by = "ne_id", tibble(ne_id = 1159137287)) %>%
```
🚫 Do not delete `1159137287` as it's a valid city near Moscow.

The action instead is the other `Noginsk` away in Siberia with **ne_id** of `1159146611` is a ghost town. We have **featurecla** = `Historic place` for that. I recommend changing the feature class instead of deleting as it does appear on most print atlases and is historically significant.  Pop values (**pop_max**, **pop_min**, **pop_other**) should go to `0`, and the pop rank (**rank_max**, **rank_min**) should go to `0`. It already has the correct **wikidata_id**.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_delete(by = "ne_id", tibble(ne_id = 1159137287)) %>%
+  #339 Gar -> Shiquanhe
+  rows_update(by = "ne_id", tibble(ne_id = 1159140703,
+                                   wikidataid = 'Q2279283')) %>%
```
In addition to the wikidataid change...

**name** to `Shiquanhe`
**name_par** to `Gar`
**name_alt** of `Ger, Sênggêkanbab, Sênggêzangbo, Sengge Zangbo, Senge Zangbo, Senge Zangbu, Sengghe Tsangpo`

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   wikidataid = 'Q2279283')) %>%
+  #334 Bol'sheretsk -> Ust-Bolsheretsk 
+  rows_update(by = "ne_id", tibble(ne_id = 1159139297,
+                                   wikidataid = 'Q2502620')) %>%
```
**name** to `Ust-Bolsheretsk`
✅  **wikidataid** to `Q2502620`
Move it north to `52.825, 156.285` like @elliotap suggested, adjusting both the lat, lng properties and the geometry.

❓ Else if that's too difficult, keep the existing feature but change the feature class and set populations to 0, and add **wikidataid** of `Q16268949`.

nvkelso reviewed (changes requested):
I'm 80% done with review (wow, there's so much good work here!), but wanted to get you early read now as there are some changes to make. I'll finish the other 20% later this week.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ISO_A2 = 'RU',
+                                   TIMEZONE = 'Europe/Moscow',
+                                   LATITUDE = 71.545556,
+                                   LONGITUDE = 52.320278)) %>%
```
Please ensure the geom lat/lng are also set.

Let's use **ne_id** = `1729769925 `.

**pop_min, pop_max, pop_other** should all be `1972` with **rank_max, rank_min** of `3`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   LONGITUDE = 52.320278)) %>%
+  #323 El Cayo -> San Ignacio 
+  rows_update(by = "ne_id", tibble(ne_id = 1159122697,
+                                   wikidataid = 'Q724815')) %>%
```
**name** to `San Ignacio`
**name_alt** to `El Cayo`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ISO_A2 = 'NO',
+                                   TIMEZONE = 'Europe/Oslo',
+                                   LATITUDE = 59.4166459,
+                                   LONGITUDE = 10.4838599)) %>%
```
🚫 Please remove this new feature for `Horten` in Norway, I closed the related issue.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   LATITUDE = 59.4166459,
+                                   LONGITUDE = 10.4838599)) %>%
+  #310 Chinhoyi should be Admin-1 capital, Kariba should not
+  rows_update(by = "ne_id", tibble(ne_id = 1159134845,
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159134845,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  rows_update(by = "ne_id", tibble(ne_id = 1159134851,
+                                   FEATURECLA = 'Populated place')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Populated place')) %>%
+  #293 Add Chad Admin-1 capitals: Fada, Pala, Massenya, Goz Beïda, Koumra
+  rows_update(by = "ne_id", tibble(ne_id = 1159148415,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+  rows_update(by = "ne_id", tibble(ne_id = 1159148415,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  rows_update(by = "ne_id", tibble(ne_id = 1159140345,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
```
👍

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ISO_A2 = 'TD',
+                                   TIMEZONE = 'Africa/Ndjamena',
+                                   LATITUDE = 11.398333,
+                                   LONGITUDE = 16.1675)) %>%
```
Also set the geom lat/lng. 

Use **ne_id** of `1729769927`.

mizmay replied:
nm ok will use this one.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ISO_A2 = 'TD',
+                                   TIMEZONE = 'Africa/Ndjamena',
+                                   LATITUDE = 12.223611,
+                                   LONGITUDE = 21.414444)) %>%
```
Also set the geom lat/lng. 

Use **ne_id** of `1729769929 `.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   ISO_A2 = 'TD',
+                                   TIMEZONE = 'Africa/Ndjamena',
+                                   LATITUDE = 8.9,
+                                   LONGITUDE = 17.55))
```
Also set the geom lat/lng. 

Use **ne_id** of `1729769933`.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  rows_update(by = "ne_id", tibble(ne_id = 1159140345,
+                                   FEATURECLA = 'Admin-1 capital')) %>%
+  rows_insert(by = "ne_id", tibble(ne_id = 1159151753,
```
Where did you get this **ne_id**, it looks old compared to newly minted Brooklyn Ints?

mizmay replied:
I was just adding new values based on max(ne_id).

Updating this one to 1729772887 , generated from Brooklyn Ints.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,187 @@
+                                   TIMEZONE = 'Africa/Ndjamena',
+                                   LATITUDE = 8.9,
+                                   LONGITUDE = 17.55))
+
```
Please also address #413 Astana, #374 Amundsen-Scott South Pole Station, and #380 Sevastopol in this PR and let's call it a day. Thank you!

mizmay replied:
Note the reason these were not updated manually is because it would be much easier to script these updates based on updates from Wikidata. The original goal of this PR was to just get the wikidataids right, and then ok fix some capitals.

mizmay replied:
Applied changes to these 4.

mizmay commented:
@nvkelso are you sure a shapefile is required here? I'm still under the assumption that the next step would be to update/add new localizations based on the wikidataids. Are you not just converting back to CSV in that step?

Updating localizations will introduce a bunch more discrepancies between NAME and name_en, which is why I didn't try to resolve anything with NAME, NAMEALT or NAMEASCII ahead of that. Some of these discrepancies help resolve issues tracked in Github issues, some are just discrepancies, some no one has raised an issue about. Step 4, see [gist](https://gist.github.com/mizmay/89f6eed0e571e5e324bb7e13005715fc).

I was thinking converting back to a shapefile, and addressing missing geometries and perhaps those to be added, would be the very last step, in a future PR.

curran commented:
Amazing!

mizmay commented:
Reran this script with @nvkelso's changes and updated the CSV.

I've also added logic to export as shapefile, however in many cases integers are being treated as double. This results in a bunch of unnecessary/false precision (e.g. 101.0000 etc.). I haven't dug deep enough to understand whether this is an issue with the data or with the r [sf](https://github.com/r-spatial/sf) library, which passes read/write commands back and fourth through GDAL.

nvkelso commented:
> I've also added logic to export as shapefile, however in many cases integers are being treated as double. This results in a bunch of unnecessary/false precision (e.g. 101.0000 etc.). I haven't dug deep enough to understand whether this is an issue with the data or with the r sf library, which passes read/write commands back and fourth through GDAL.

When I've done similar in Python using Shapely I have to ready the field definitions from the input file and then copy that (with modification if necessary) for the output file, else it'll make wrong guesses like ints and doubles. Is there a similar convention for the R library?

mizmay commented:
@nvkelso there are ways to do this. If you have a moment, provide a link for where field type definitions are defined?

nvkelso commented:
> @nvkelso there are ways to do this. If you have a moment, provide a link for where field type definitions are explicitly defined?

I'd match what `ogrinfo` says.

mizmay commented:
Problem with this is that the field type definitions for this shapefile are suspicious -- there are a bunch of fields defined as real numbers with 11 decimal places of precision, and the GDAL Shapefile driver handles this by adding zeros. See ogrinfo output below.

What I did instead was to find all the numeric fields I think *should* be integers (by testing whether they contain whole numbers), and converted those to integer fields.

Here are the fields I converted to integer if they weren't already, based on a count of values in the fields that were integers:
```
   fields     integer_values
   <chr>               <int>
 1 SCALERANK            7339
 2 NATSCALE             7339
 3 LABELRANK            7339
 4 DIFFASCII            7339
 5 ADM0CAP              7339
 6 WORLDCITY            7339
 7 MEGACITY             7339
 8 CHANGED              7339
 9 NAMEDIFF             7339
10 POP_MAX              7341
11 POP_MIN              7340
12 POP_OTHER            7340
13 RANK_MAX             7341
14 RANK_MIN             7340
15 GEONAMEID            7311
16 LS_MATCH             7339
17 CHECKME              7339
18 MAX_POP10            7339
19 MAX_POP20            7339
20 MAX_POP50            7339
21 MAX_POP300           7339
22 MAX_POP310           7339
23 MAX_NATSCA           7339
24 MIN_AREAKM           7339
25 MAX_AREAKM           7339
26 MIN_AREAMI           7339
27 MAX_AREAMI           7339
28 MIN_PERKM            7339
29 MAX_PERKM            7339
30 MIN_PERMI            7339
31 MAX_PERMI            7339
32 COMPARE              7339
33 ADMIN1_COD           7339
34 GN_POP               7339
35 ELEVATION            7339
36 GTOPO30              7339
37 UN_FID               7339
38 POP1950              7339
39 POP1955              7339
40 POP1960              7339
41 POP1965              7339
42 POP1970              7339
43 POP1975              7339
44 POP1980              7339
45 POP1985              7339
46 POP1990              7339
47 POP1995              7339
48 POP2000              7339
49 POP2005              7339
50 POP2010              7339
51 POP2015              7339
52 POP2020              7339
53 POP2025              7339
54 POP2050              7339
55 wof_id               7339
56 wdid_score           7163
57 ne_id                7344
```

OGRINFO (field size and precision in parentheses)

`ogrinfo -al ne_10m_populated_places.shp > ne_10m_populated_places.meta`

```
SCALERANK: Integer (4.0)
NATSCALE: Integer (4.0)
LABELRANK: Integer (4.0)
FEATURECLA: String (50.0)
NAME: String (100.0)
NAMEPAR: String (254.0)
NAMEALT: String (254.0)
DIFFASCII: Integer (4.0)
NAMEASCII: String (100.0)
ADM0CAP: Real (19.11)
CAPIN: String (15.0)
WORLDCITY: Real (19.11)
MEGACITY: Integer (4.0)
SOV0NAME: String (100.0)
SOV_A3: String (3.0)
ADM0NAME: String (50.0)
ADM0_A3: String (3.0)
ADM1NAME: String (50.0)
ISO_A2: String (5.0)
NOTE: String (254.0)
LATITUDE: Real (19.11)
LONGITUDE: Real (19.11)
CHANGED: Real (19.11)
NAMEDIFF: Integer (4.0)
DIFFNOTE: String (254.0)
POP_MAX: Integer (9.0)
POP_MIN: Integer (9.0)
POP_OTHER: Integer (9.0)
RANK_MAX: Integer (9.0)
RANK_MIN: Integer (9.0)
GEONAMEID: Real (19.11)
MEGANAME: String (50.0)
LS_NAME: String (41.0)
LS_MATCH: Integer (4.0)
CHECKME: Integer (4.0)
MAX_POP10: Real (19.0)
MAX_POP20: Real (19.0)
MAX_POP50: Real (19.0)
MAX_POP300: Real (19.0)
MAX_POP310: Real (19.0)
MAX_NATSCA: Real (19.0)
MIN_AREAKM: Real (19.0)
MAX_AREAKM: Real (19.11)
MIN_AREAMI: Real (19.11)
MAX_AREAMI: Real (19.11)
MIN_PERKM: Real (19.11)
MAX_PERKM: Real (19.11)
MIN_PERMI: Real (19.11)
MAX_PERMI: Real (19.11)
MIN_BBXMIN: Real (19.11)
MAX_BBXMIN: Real (19.11)
MIN_BBXMAX: Real (19.11)
MAX_BBXMAX: Real (19.11)
MIN_BBYMIN: Real (19.11)
MAX_BBYMIN: Real (19.11)
MIN_BBYMAX: Real (19.11)
MAX_BBYMAX: Real (19.11)
MEAN_BBXC: Real (19.11)
MEAN_BBYC: Real (19.11)
COMPARE: Integer (4.0)
GN_ASCII: String (254.0)
FEATURE_CL: String (254.0)
FEATURE_CO: String (254.0)
ADMIN1_COD: Real (19.11)
GN_POP: Real (19.0)
ELEVATION: Real (19.11)
GTOPO30: Real (19.11)
TIMEZONE: String (254.0)
GEONAMESNO: String (100.0)
UN_FID: Integer (9.0)
UN_ADM0: String (254.0)
UN_LAT: Real (19.11)
UN_LONG: Real (19.11)
POP1950: Real (19.11)
POP1955: Real (19.11)
POP1960: Real (19.11)
POP1965: Real (19.11)
POP1970: Real (19.11)
POP1975: Real (19.11)
POP1980: Real (19.11)
POP1985: Real (19.11)
POP1990: Real (19.11)
POP1995: Real (19.11)
POP2000: Real (19.11)
POP2005: Real (19.11)
POP2010: Real (19.11)
POP2015: Real (19.11)
POP2020: Real (19.11)
POP2025: Real (19.11)
POP2050: Real (19.11)
CITYALT: String (50.0)
min_zoom: Real (6.1)
wikidataid: String (30.0)
wof_id: Integer64 (12.0)
CAPALT: Integer (4.0)
name_en: String (254.0)
name_de: String (254.0)
name_es: String (254.0)
name_fr: String (254.0)
name_pt: String (254.0)
name_ru: String (254.0)
name_zh: String (254.0)
label: String (254.0)
name_ar: String (254.0)
name_bn: String (254.0)
name_el: String (254.0)
name_hi: String (254.0)
name_hu: String (254.0)
name_id: String (254.0)
name_it: String (254.0)
name_ja: String (254.0)
name_ko: String (254.0)
name_nl: String (254.0)
name_pl: String (254.0)
name_sv: String (254.0)
name_tr: String (254.0)
name_vi: String (254.0)
wdid_score: Integer (1.0)
ne_id: Integer64 (10.0)
```

mizmay commented:
Note this does result in field type definition and precision differences from the original: SCALERANK: Integer (9.0):

Results exported and included in this PR as `ne_10m_populated_places_modified.shp`

```
NATSCALE: Integer (9.0)
LABELRANK: Integer (9.0)
FEATURECLA: String (80.0)
NAME: String (80.0)
NAMEPAR: String (80.0)
NAMEALT: String (95.0)
DIFFASCII: Integer (9.0)
NAMEASCII: String (80.0)
ADM0CAP: Integer (9.0)
CAPIN: String (80.0)
WORLDCITY: Integer (9.0)
MEGACITY: Integer (9.0)
SOV0NAME: String (80.0)
SOV_A3: String (80.0)
ADM0NAME: String (80.0)
ADM0_A3: String (80.0)
ADM1NAME: String (80.0)
ISO_A2: String (80.0)
NOTE: String (80.0)
LATITUDE: Real (24.15)
LONGITUDE: Real (24.15)
CHANGED: Integer (9.0)
NAMEDIFF: Integer (9.0)
DIFFNOTE: String (99.0)
POP_MAX: Integer (9.0)
POP_MIN: Integer (9.0)
POP_OTHER: Integer (9.0)
RANK_MAX: Integer (9.0)
RANK_MIN: Integer (9.0)
GEONAMEID: Integer (9.0)
MEGANAME: String (80.0)
LS_NAME: String (80.0)
LS_MATCH: Integer (9.0)
CHECKME: Integer (9.0)
MAX_POP10: Integer (9.0)
MAX_POP20: Integer (9.0)
MAX_POP50: Integer (9.0)
MAX_POP300: Integer (9.0)
MAX_POP310: Integer (9.0)
MAX_NATSCA: Integer (9.0)
MIN_AREAKM: Integer (9.0)
MAX_AREAKM: Integer (9.0)
MIN_AREAMI: Integer (9.0)
MAX_AREAMI: Integer (9.0)
MIN_PERKM: Integer (9.0)
MAX_PERKM: Integer (9.0)
MIN_PERMI: Integer (9.0)
MAX_PERMI: Integer (9.0)
MIN_BBXMIN: Real (24.15)
MAX_BBXMIN: Real (24.15)
MIN_BBXMAX: Real (24.15)
MAX_BBXMAX: Real (24.15)
MIN_BBYMIN: Real (24.15)
MAX_BBYMIN: Real (24.15)
MIN_BBYMAX: Real (24.15)
MAX_BBYMAX: Real (24.15)
MEAN_BBXC: Real (24.15)
MEAN_BBYC: Real (24.15)
COMPARE: Integer (9.0)
GN_ASCII: String (80.0)
FEATURE_CL: String (80.0)
FEATURE_CO: String (80.0)
ADMIN1_COD: Integer (9.0)
GN_POP: Integer (9.0)
ELEVATION: Integer (9.0)
GTOPO30: Integer (9.0)
TIMEZONE: String (80.0)
GEONAMESNO: String (100.0)
UN_FID: Integer (9.0)
UN_ADM0: String (80.0)
UN_LAT: Real (24.15)
UN_LONG: Real (24.15)
POP1950: Integer (9.0)
POP1955: Integer (9.0)
POP1960: Integer (9.0)
POP1965: Integer (9.0)
POP1970: Integer (9.0)
POP1975: Integer (9.0)
POP1980: Integer (9.0)
POP1985: Integer (9.0)
POP1990: Integer (9.0)
POP1995: Integer (9.0)
POP2000: Integer (9.0)
POP2005: Integer (9.0)
POP2010: Integer (9.0)
POP2015: Integer (9.0)
POP2020: Integer (9.0)
POP2025: Integer (9.0)
POP2050: Integer (9.0)
CITYALT: String (80.0)
min_zoom: Real (24.15)
wikidataid: String (80.0)
wof_id: Integer64 (10.0)
CAPALT: Integer (9.0)
name_en: String (80.0)
name_de: String (80.0)
name_es: String (80.0)
name_fr: String (80.0)
name_pt: String (80.0)
name_ru: String (80.0)
name_zh: String (80.0)
label: String (80.0)
name_ar: String (80.0)
name_bn: String (80.0)
name_el: String (80.0)
name_hi: String (80.0)
name_hu: String (80.0)
name_id: String (80.0)
name_it: String (80.0)
name_ja: String (80.0)
name_ko: String (80.0)
name_nl: String (80.0)
name_pl: String (80.0)
name_sv: String (80.0)
name_tr: String (80.0)
name_vi: String (80.0)
wdid_score: Integer (9.0)
```

nvkelso commented:
Hey look, fixing old issues left and right! https://github.com/nvkelso/natural-earth-vector/issues/5

nvkelso commented:
I reviewed existing values in the populated places theme, and associated themes that drive values in some of the columns. The table below as the `old` (original) field definition, the `draft` in this PR as of the last commit, and my `proposed` definition based on this research. I explain some rational in the `comment`.

In all cases I'm using 100 char for names.

 In some cases I'm marking some fields with `¿Deprecate?` in the comment but NOT in the proposed – this PR shouldn't delete any columns. 

property | old | draft | proposed | comment
--- | --- | --- | --- | ---
SCALERANK |  Integer (4.0) |  Integer (9.0) |  Integer (2.0) | Values are 0-10
NATSCALE |  Integer (4.0) |  Integer (9.0) |  Integer (4.0) | Values are 1-600
LABELRANK |  Integer (4.0) |  Integer (9.0) |  Integer (2.0) | Values are 0-10
FEATURECLA |  String (50.0) |  String (80.0) |  String (50.0) | String length 30 to match many other themes (sometimes 50)
NAME |  String (100.0) |  String (80.0) |  String (100.0) | Length should be 100 (to match many other themes)
NAMEPAR |  String (254.0) |  String (80.0) |  String (100.0) | Length should be 100 (to allow single parenthetical value)
NAMEALT |  String (254.0) |  String (95.0) |  String (254.0) | Length should be 254 to allow multiple alt values, comma delim
DIFFASCII |  Integer (4.0) |  Integer (9.0) |  Integer (4.0) | ¿Deprecate?
NAMEASCII |  String (100.0) |  String (80.0) |  String (100.0) | Length should be 100 (to match name)
ADM0CAP |  Real (19.11) |  Integer (9.0) |  Integer (1.0) | Int (boolean)
CAPIN |  String (15.0) |  String (80.0) |  String (100.0) | ¿String should be longer, like 100 (some values cut off now)
WORLDCITY |  Real (19.11) |  Integer (9.0) |  Integer (1.0) | Int (boolean) – 70 count
MEGACITY |  Integer (4.0) |  Integer (9.0) |  Integer (1.0) | Int (boolean) – 590 count
SOV0NAME |  String (100.0) |  String (80.0) |  String (100.0) | ¿What is length in admin layers? 32
SOV_A3 |  String (3.0) |  String (80.0) |  String (3.0) | String length 3 (because it's a NE-ism)
ADM0NAME |  String (50.0) |  String (80.0) |  String (100.0) | ¿What is length in admin layers? 36
ADM0_A3 |  String (3.0) |  String (80.0) |  String (3.0) | String length 3 (because it's a NE-ism)
ADM1NAME |  String (50.0) |  String (80.0) |  String (100.0) | ¿What is length in admin layers? 44 to 129. Do 100
ISO_A2 |  String (5.0) |  String (80.0) |  String (2.0) | String length 2
NOTE |  String (254.0) |  String (80.0) |  String (254.0) | ¿String should be 254?
LATITUDE |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
LONGITUDE |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
CHANGED |  Real (19.11) |  Integer (9.0) |  Integer (1.0) | ¿Deprecate?
NAMEDIFF |  Integer (4.0) |  Integer (9.0) |  Integer (1.0) | ¿Deprecate?
DIFFNOTE |  String (254.0) |  String (99.0) |  String (254.0) | ¿Deprecate? Else string 255
POP_MAX |  Integer (9.0) |  Integer (9.0) |  Integer (12.0) | Int, largest value would be 10B, so 10x that 100000000000 is 12?
POP_MIN |  Integer (9.0) |  Integer (9.0) |  Integer (12.0) | Int
POP_OTHER |  Integer (9.0) |  Integer (9.0) |  Integer (12.0) | Int
RANK_MAX |  Integer (9.0) |  Integer (9.0) |  Integer (2.0) | Int
RANK_MIN |  Integer (9.0) |  Integer (9.0) |  Integer (2.0) | Int
GEONAMEID |  Real (19.11) |  Integer (9.0) |  Integer (9.0) | Int
MEGANAME |  String (50.0) |  String (80.0) |  String (100.0) | String
LS_NAME |  String (41.0) |  String (80.0) |  String (100.0) | ¿String length 50? What is it in landscan theme – 41.
LS_MATCH |  Integer (4.0) |  Integer (9.0) |  Integer (4.0) | ¿Deprecate?
CHECKME |  Integer (4.0) |  Integer (9.0) |  Integer (4.0) | ¿Deprecate?
MAX_POP10 |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_POP20 |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_POP50 |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_POP300 |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_POP310 |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_NATSCA |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MIN_AREAKM |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | Int
MAX_AREAKM |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MIN_AREAMI |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MAX_AREAMI |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MIN_PERKM |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MAX_PERKM |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MIN_PERMI |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MAX_PERMI |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
MIN_BBXMIN |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MAX_BBXMIN |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MIN_BBXMAX |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MAX_BBXMAX |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MIN_BBYMIN |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MAX_BBYMIN |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MIN_BBYMAX |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MAX_BBYMAX |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MEAN_BBXC |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
MEAN_BBYC |  Real (19.11) |  Real (24.15) |  Real (9.6) | This should be signed 3 digits plus 6 precision
COMPARE |  Integer (4.0) |  Integer (9.0) |  Integer (1.0) | ¿Deprecate?
GN_ASCII |  String (254.0) |  String (80.0) |  String (254.0) | ¿Deprecate? Compare with name and name_alts first
FEATURE_CL |  String (254.0) |  String (80.0) |  String (254.0) | ¿Deprecate?
FEATURE_CO |  String (254.0) |  String (80.0) |  String (254.0) | ¿Deprecate? Look at values first to see if there are more adm-2 caps to harvest?
ADMIN1_COD |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | ¿Deprecate?
GN_POP |  Real (19.0) |  Integer (9.0) |  Integer (12.0) | ¿Deprecate? Compare with populations first?
ELEVATION |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | ¿Deprecate? Or Int and move GTOPO30 here when null?
GTOPO30 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | ¿Deprecate?
TIMEZONE |  String (254.0) |  String (80.0) |  String (50.0) | String 32 chars?
GEONAMESNO |  String (100.0) |  String (100.0) |  String (100.0) | ¿Deprecate?
UN_FID |  Integer (9.0) |  Integer (9.0) |  Integer (9.0) | Int
UN_ADM0 |  String (254.0) |  String (80.0) |  String (100.0) | ¿Deprecate?
UN_LAT |  Real (19.11) |  Real (24.15) |  Real (9.6) | ¿Deprecate?
UN_LONG |  Real (19.11) |  Real (24.15) |  Real (9.6) | ¿Deprecate?
POP1950 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1955 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1960 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1965 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1970 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1975 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1980 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1985 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1990 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP1995 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2000 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2005 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2010 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2015 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2020 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2025 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
POP2050 |  Real (19.11) |  Integer (9.0) |  Integer (12.0) | Int
CITYALT |  String (50.0) |  String (80.0) |  String (100.0) | ¿Deprecate? Compare with name and name_alts first
min_zoom |  Real (6.1) |  Real (24.15) |  Real (3.1) | Only needs 1 decimal precision
wikidataid |  String (30.0) |  String (80.0) |  String (30.0) | String
wof_id |  Integer64 (12.0) |  Integer64 (10.0) |  Integer64 (12.0) | Int
CAPALT |  Integer (4.0) |  Integer (9.0) |  Integer (1.0) | Int (boolean)
name_en |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_de |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_es |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_fr |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_pt |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_ru |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_zh |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
label |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_ar |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_bn |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_el |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_hi |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_hu |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_id |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_it |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_ja |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_ko |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_nl |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_pl |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_sv |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_tr |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
name_vi |  String (254.0) |  String (80.0) |  String (100.0) | String, length should match name column
wdid_score |  Integer (1.0) |  Integer (9.0) |  Integer (1.0) | ¿Deprecate?
ne_id |  Integer64 (10.0) |  Integer64 (10.0) |  Integer64 (12.0) | Int

nvkelso commented:
I was going to edit the PR to use my proposed SHP/DBF field definitions but I don't see where to do that. @mizmay can you point the way, please?

mizmay commented:
I don't think there is any GDAL method that would allow you to explicitly declare or ingest a mapping for field size and precision when writing to shapefiles. https://gdal.org/drivers/vector/shapefile.html#vector-shapefile

I believe there is a Python binding that may enable what you are remembering - https://pcjericks.github.io/py-gdalogr-cookbook/vector_layers.html

Good luck!

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,270 @@
+# should verify this doesn't introduce regressions.
+# If not, it's the most straightforward strategy.
+ne_10m_populated_places_modified <- st_as_sf(ne_10m_populated_places_modified_metadata,
+                                             coords = c("LONGITUDE","LATITUDE"), remove = FALSE,
```
NEW: At the very top... can you set the `"LONGITUDE","LATITUDE"` properties to the geometry values... 

Then in the middle of the script you modify the `"LONGITUDE","LATITUDE"` properties (no change needed there).

Then at the bottom you set the geom to the new values to make them consistent (no change needed).

mizmay replied:
This is done. I've included a csv in the output that lists anyplace where this changed values:
1 - a change that won't be visible/meaningful (600 of these)
2 - a change that will be visible (16 of these, nothing I can see that should hold up this PR)

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,270 @@
+  st_drop_geometry %>%
+  summarise(across(where(is.numeric), count.int)) %>%
+  pivot_longer(everything(), names_to = "fields", values_to = "integer_values") %>%
+  filter(integer_values > 7000)
```
What is this magic number `7000`?

mizmay replied:
This was laziness to get around places where there were a few missing values - fixed it so that fields are classed as integers if all values are either integers or missing.

nvkelso commented on updates/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,270 @@
+  mutate(across(all_of(counts.int$fields), as.integer, na.rm = TRUE))
+
+
+st_write(ne_10m_populated_places_modified, "ne_10m_populated_places_modified.shp",
```
This output DBF just needs a sanity check for char encoding and to see what fields were actually changed, and properties from the PR.

Optional: in QGIS redo all the fields to match the issue comment.

mizmay replied:
<img width="1166" alt="Screen Shot 2021-03-03 at 7 58 42 PM" src="https://user-images.githubusercontent.com/3074691/109909390-e113bc80-7c5a-11eb-8772-beb603fb5e26.png">

Encoding looks fine, see Goz Beïda in screenshot above for one example. There are lots of missing values for the new records, e.g. SCALERANK, maybe best assigned manually in QGIS?

nvkelso commented:
Please move the R script the output SHP to a new dir in tools:

> tools/457_fix_wikidataids

nvkelso commented:
I loaded the non-spatial DBFs into OpenOffice and copied over the key/value table contents to text files (tab delim) and compared them in BBEdit. 

As expected the field (column) formats are different in the two files and I'll touch those up to match the proposal in https://github.com/nvkelso/natural-earth-vector/issues/5. None of the values changed, so no data loss just formatting decorations.

The values in the LATITUDE and LONGITUDE columns were all clipped to 6 significant digits as expected. I then removed those 2 columns from the compare to simplify life.

That left the several dozen rows with the changes expected from this PR. I spot checked those and they all look as expected.

I will merge this PR, apply some edits to the SHP/DBF, copy it over the 10m_cultural theme to make it ready for release.

Thank you @mizmay for the huge assist!

nvkelso commented on tools/457_fix_wikidataids/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,295 @@
+                                   RANK_MAX = 11)) %>%
+  #407 Add Nnewi, Nigeria
+  rows_insert(by = "ne_id", tibble(ne_id = 1729764357,
+                                   wikidataid = 'Q2750772',
```
This should have set NAME and NAMEASCII to `Nnewi`

nvkelso commented on tools/457_fix_wikidataids/Fix_10m_Populated_Places_WikidataIDs.R:
```diff
@@ -0,0 +1,295 @@
+                                   LONGITUDE = 156.285)) %>%
+  #330 Add Belushya Guba, Russia
+  rows_insert(by = "ne_id", tibble(ne_id = 1729769925,
+                                   wikidataid = 'Q26324',
```
This should have set NAME and NAMEASCII to `Belushya Guba`.

nvkelso commented:
As noted in conversation, I later ensured each NEW record had the minimum properties:

- `NE_ID` - from [Brooklyn Integers](https://www.brooklynintegers.com/create/)
- `NAME` - based on issue decision
- `NAMEASCII` - without any diacritical marks
- `NAMEPAR` - only when single other name in common usage
- `NAMEALT` - when appropriate, a list
- `SCALERANK` - look around at similarly sized cities in that map area
- `NATSCALE` - look around at similarly sized cities in that map area
- `LABELRANK` - look around at similarly sized cities in that map area
- `FEATURECLA` - based on issue decision or no change
- `ADM0CAP` - set to 1 if a country capital, else 0
- `CAPIN` - only if it's a country capital
- `WORLDCITY` - to 0 since those are controlled lists
- `MEGACITY` - to 0 since those are controlled lists
- `SOV0NAME` - based on what NE sov it falls in
- `SOV_A3` - based on what NE sov it falls in
- `ADM0NAME` - based on what NE country it falls in
- `ADM0_A3` - based on what NE country it falls in
- `ADM1NAME` - based on what NE admin1 it falls in, or it's real world value
- `ISO_A2` - based on what NE country it falls in
- `POP_MAX` - these are needed to set townspot size and label text size
- `POP_MIN` - these are needed to set townspot size and label text size
- `RANK_MAX` - these are needed to set townspot size and label text size
- `RANK_MIN` - these are needed to set townspot size and label text size
- `GEONAMEID` - when available
- `MIN_ZOOM` - look around at similarly sized cities in that map area
- `WIKIDATAID` - when available
- `WOF_ID` - when available

NOTE: all the `name_*` localizations are picked up later by running the scripts in https://github.com/nvkelso/natural-earth-vector/tree/master/tools/wikidata.
